PopPUNK/network.py

```diff
@@ -399,33 +469,74 @@ def constructNetwork(rlist, qlist, assignments, within_label,
                                                                                 weights = True)
             for (ref, query, weight) in zip(extra_sources, extra_targets, extra_weights):
                 edge_tuple = (ref, query, weight)
-                connections.append(edge_tuple)
+                if ref < query:
+                    connections.append(edge_tuple)
         else:
             extra_sources, extra_targets = load_previous_network(prev_G,rlist,
                                                                                 weights = False)
             for (ref, query) in zip(extra_sources, extra_targets):
                 edge_tuple = (ref, query)
-                connections.append(edge_tuple)
-
-    # build the graph
-    G = gt.Graph(directed = False)
-    G.add_vertex(len(vertex_labels))
+                if ref < query:
+                    connections.append(edge_tuple)
+
+    # load GPU libraries if necessary
+    if use_gpu:
+        
+        # load CUDA libraries
+        try:
+            import cugraph
+            import cudf
+        except ImportError as e:
+            sys.stderr.write("cugraph and cudf unavailable\n")
+            raise ImportError(e)
+            
+        # create DataFrame using edge tuples
+        if weights is not None or sparse_input is not None:
+            G_df = cudf.DataFrame(connections, columns =['source', 'destination', 'weights'])
+        else:
+            G_df = cudf.DataFrame(connections, columns =['source', 'destination'])
+        
+        # ensure the highest-integer node is included in the edge list
+        # by adding a self-loop if necessary; see https://github.com/rapidsai/cugraph/issues/1206
+        max_in_df = np.amax([G_df['source'].max(),G_df['destination'].max()])
+        max_in_vertex_labels = len(vertex_labels)-1
+        if max_in_df.item() != max_in_vertex_labels:
+            G_self_loop = cudf.DataFrame()
+            G_self_loop['source'] = [max_in_vertex_labels]
+            G_self_loop['destination'] = [max_in_vertex_labels]
+            if weights is not None or sparse_input is not None:
+                G_self_loop['weights'] = [0.0]
+            G_df = cudf.concat([G_df,G_self_loop], ignore_index = True)
+            new_max_in_df = np.amax([G_df['source'].max(),G_df['destination'].max()])
```

Comment:
This looks the same as some code in the function above? So, it should be its own function so it's not copied

Reply:
Function created in f61ccd0.